src/qat/model/hardware_model.py

```diff
@@ -1,48 +1,124 @@
-    physical_channels: ComponentDict[PhysicalChannel] = Field(frozen=True, default=dict())
-    pulse_channels: ComponentDict[PulseChannel] = Field(frozen=True, default=dict())
-    qubits: ComponentDict[Qubit] = Field(frozen=True, default=dict())
-    resonators: ComponentDict[Resonator] = Field(frozen=True, default=dict())
+    version: SemanticVersion = Field(frozen=True, repr=False, default=VERSION)
+    topology: dict[QubitId, set[QubitId]]
 
     @field_validator("version")
-    def version_compatibility(version):
+    def version_compatibility(version: Version):
         assert (
             version.major == VERSION.major
         ), f"Direct instantiation requires major version compatibility (expected {VERSION.major}.Y.Z, found {version})"
         assert (
             VERSION >= version
         ), f"Latest supported hardware model version {VERSION}, found {version}"
         return VERSION
+
+    def __eq__(self, other: LogicalHardwareModel) -> bool:
+        if type(self) != type(other):
+            return False
+
+        if self.model_fields != other.model_fields:
+            return False
+
+        if self.version != other.version:
+            return False
+
+        if self.topology != other.topology:
+            return False
+
+        return True
+
+    def __ne__(self, other: LogicalHardwareModel) -> bool:
+        return not self.__eq__(other)
+
+
+class QuantumHardwareModel(LogicalHardwareModel):
+    """Class for calibrating our QPU hardware.
+
+    Attributes:
+        qubits: The superconducting qubits on the chip.
+        constrained_topology: The connectivities between the qubits, which can be a subgraph of the physical topology.
+    """
+
+    qubits: dict[QubitId, Qubit]
+    constrained_topology: Optional[dict[QubitId, set[QubitId]]] = None
+
+    @model_validator(mode="before")
+    def validate_topology(cls, data):
+        topology = data["topology"]
+
+        try:
+            constrained_topology = data["constrained_topology"]
+
+            for qubit_index in topology:
+                if not constrained_topology[qubit_index] <= topology[qubit_index]:
+                    raise ValueError(
+                        "Constrained topology must be a subgraph of the physical topology."
+                    )
+        except (KeyError, TypeError):
+            data["constrained_topology"] = topology
+
+        return data
+
+    @field_validator("topology")
+    def validate_topology_symmetry(cls, topology):
+        for node, connected_nodes in topology.items():
+            for connected_node in connected_nodes:
+                if node not in topology[connected_node]:
+                    raise ValueError(
+                        f"The topology is not symmetric, node {node} not present in connected nodes of node {connected_node}."
+                    )
+        return topology
+
+    @field_validator("constrained_topology")
+    def validate_constrained_topology_symmetry(cls, constrained_topology):
```

Comment:
Does the topology actually need to be symmetric here?

For example, a particular hardware model might directly support a CNOT with control qubit 4 and target qubit 5, but not (directly) support a CNOT with control qubit 5 and target qubit 4. Of course in practice we can always indirectly achieve the CNOT with the opposite direction by use of Hadamard gates.

My thoughts are if this wasn't required to be symmetric, this could, for example, guide the compilation in determining how to implement the CNOT in the opposite direction (for this example).

Reply:
I honestly have no idea! You could be right, I'll ask around to give you a more definite answer on this.

Reply:
OK, I think we need to remove this symmetric validator.